### PR TITLE
Cockpit chatbot

### PR DESCRIPTION
**Pull request by abdulrahmantkhalifa**
relates to #245

### CODE REVIEW

File: ays_bot/telegrambot/Repo.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        self.reg_repo = re.compile(r'^[a-zA-Z0-9-_]+$')
+
+    def init_repo(self, update, bot, name, git_url):
+        chat_id = update.message.chat_id
+        bot.sendChatAction(chat_id=chat_id, action=telegram.ChatAction.TYPING)
+
+        repopath = '%s/%s' % (self.rootpath, name)
+        self.bot.logger.debug('initializing repository: %s' % repopath)
+
+        j.atyourservice.repoCreate(repopath)
```
Comment (zaibon) [inline ays_bot/telegrambot/Repo.py]: `repoCreate` taks the `git_url` as second argument. no need to manually to create the git repo and add the remote to it.

File: ays_bot/telegrambot/Repo.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+        if self._cuisine.core.run("cd %s && git remote -v" % repopath)[1]:
+            self._cuisine.core.run("cd %s && git remote set-url origin  %s " % (repopath, git_url))
+        else:
+            self._cuisine.core.run("cd %s && git remote add origin  %s " % (repopath, git_url))
+        evt = j.data.models.cockpit_event.Telegram()
```
Comment (zaibon) [inline ays_bot/telegrambot/Repo.py]: don't need to send an event for this anymore.
Aslo event are not handle like this.

File: ays_bot/telegrambot/Repo.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        username = update.message.from_user.username
+        bot.sendChatAction(chat_id=chat_id, action=telegram.ChatAction.TYPING)
+
+        # check repo name validity
+        if not re.search(self.reg_repo, repo_name):
+            message = "Sorry, I don't support this repo name, please name it without any special characters or spaces."
+            return self.bot.sendMessage(chat_id=chat_id, text=message)
+
+        # repo already exists
+        path = j.sal.fs.joinPaths(j.dirs.codeDir, 'cockpit', repo_name)
```
Comment (zaibon) [inline ays_bot/telegrambot/Repo.py]: to test if a repo already exists check is presence in `j.atyourservice.reposList`

File: ays_bot/telegrambot/Repo.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+        if not self._currentRepoName(username):
+            self.bot.sendMessage(chat_id=chat_id, text="No repo selected now.")
+        else:
+            self.bot.sendMessage(
+                chat_id=chat_id,
+                text="Current repo: *%s*" %
+                self._currentRepoName(username),
+                parse_mode=telegram.ParseMode.MARKDOWN)
+
+        repos = j.atyourservice.reposDiscover(AYS_REPO_DIR)
```
Comment (zaibon) [inline ays_bot/telegrambot/Repo.py]: `reposDiscover` is a pretty heavy action, prefer the `j.atyourservice.reposList` instead

File: ays_bot/telegrambot/Repo.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+                    text=message,
+                    parse_mode=telegram.ParseMode.MARKDOWN,
+                    reply_markup=reply_markup)
+                continue
+
+            if name == self._currentRepoName(username):
+                self._setCurrentRepo(username, None)
+
+            self.bot.logger.debug('removing repository: %s' % repo.path)
+            j.sal.fs.removeDirTree(repo.path)
```
Comment (zaibon) [inline ays_bot/telegrambot/Repo.py]: to delete a repository properly get the repo object then call destroy on it. just deleting it from the filesystem doesn't remove it from the database

File: ays_bot/telegrambot/Repo.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+
+            evt = j.data.models.cockpit_event.Telegram()
```
Comment (zaibon) [inline ays_bot/telegrambot/Repo.py]: no need to send an event for this anymore

File: ays_bot/telegrambot/Service.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        services_list = self._servicelist(repo)
+
+        msg = '\n'.join(services_list).replace("_", "\_")
+        self.bot.sendMessage(chat_id=update.message.chat_id, text=msg, parse_mode=telegram.ParseMode.MARKDOWN)
+
+    def inspect(self, bot, update, service_id):
+        username = update.message.from_user.username
+        repo = self._currentRepo(username)
+        service = repo.serviceGetByKey(service_id).model
+        msg = """Role: *{role}*
```
Comment (zaibon) [inline ays_bot/telegrambot/Service.py]: We need a better view for service.
HRD is not relevant name since we don't use HRD anymore. data or schema is more appropriate.

For the state we probably want to display the state of all the actions of the service
`service.model.actionsState`

Showing it's parent and producers can also be usefull information.
Have a look at the `ays show` command to see what information are relevant.

File: ays_bot/telegrambot/Service.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+            text=msg,
+            parse_mode=telegram.ParseMode.MARKDOWN,
+            reply_markup=telegram.ReplyKeyboardHide())
+
+    def delete(self, bot, update, service_id):
+        username = update.message.from_user.username
+        repo = self._currentRepo(username)
+        service = repo.serviceGetByKey(service_id).model
+        try:
+            service.delete()
```
Comment (zaibon) [inline ays_bot/telegrambot/Service.py]: Here you call delete only on the model of the service. you should call delete on the service object itself. it has more logic inside.

File: ays_bot/telegrambot/Service.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+            msg = "serice hase been deleted successfully"
+        except Exception as e:
+            self.bot.logger("an Error occured during delete of service : %s" % str(e))
+            msg = "serice failed to deleted, service might be in use.\n if error presists please contact our support team."
+        bot.sendMessage(text=msg,
+                        chat_id=chat_id)
+
+    def choose_action(self, bot, update):
+        self.callbacks[update.message.from_user.username] = self.dispatch_choice
+        choices = ['list', 'inspect', 'execute']
```
Comment (zaibon) [inline ays_bot/telegrambot/Service.py]: remove execute from this list since we don't provide that action.

File: ays_bot/telegrambot/run.py [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+        return self.bot.repo_mgmt._currentRepoName(username)
+
+    def _currentRepo(self, username):
+        return j.atyourservice.repoGet(j.sal.fs.joinPaths(AYS_REPO_DIR, self._currentRepoName(username)))
+
+    def _createname(self, run):
+        return datetime.datetime.fromtimestamp(run.dictFiltered["lastModDate"]).strftime('%c')
+
+    def _runlist(self, username):
+        repo = self._currentRepo(username)
```
Comment (zaibon) [inline ays_bot/telegrambot/run.py]: to list all runs from a repo use `repo.runsList()`

File: ays_bot/telegrambot/run.py [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+            text=run.__str__())
+
+    def execute(self, bot, update, run_id):
+
+        username = update.message.from_user.username
+        chat_id = update.message.chat_id
+        repo = self._currentRepo(username)
+        run = repo.runGet(run_id).objectGet()
+        try:
+            run.execute()
```
Comment (zaibon) [inline ays_bot/telegrambot/run.py]: We can't execute the run synchronously here. This will block the telegram bot for too long.
Instead we need to execute it through the AYS daemon asynchronously. Check `j.clients.atyourservice` to see how to schedule run to the AYS daemon.

File: ays_bot/telegrambot/run.py [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+
+        msg = '\n'.join(runs_list)
+        self.bot.sendMessage(chat_id=update.message.chat_id, text=msg, parse_mode=telegram.ParseMode.MARKDOWN)
+
+    def inspect(self, bot, update, run_id):
+        username = update.message.from_user.username
+        repo = self._currentRepo(username)
+        run = repo.runGet(run_id)
+        # data = ''
+        data = j.data.serializer.yaml.dumps(run.dictFiltered)
```
Comment (zaibon) [inline ays_bot/telegrambot/run.py]: I don't think that''s a good enough format for end user

File: ays_bot/telegrambot/templates/blueprint [hunk 13]
```diff
@@ -0,0 +1,18 @@
+g8client__main:
```
Comment (zaibon) [inline ays_bot/telegrambot/templates/blueprint]: Why is this here ?

File: ays_bot/telegrambot/Bot.py [hunk 14]
```diff
@@ -0,0 +1,422 @@
...
+class TGBot():
+
+    # initializing
+    def __init__(self, config, rootpath='', redis=None):
+        self.token = config.get('token', '')
+        self.config = config
+        self._errmessages = []
+
+        if rootpath == "":
+            rootpath = j.sal.fs.joinPaths(j.dirs.codeDir, "cockpit")
```
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/Bot.py]: done

File: ays_bot/telegrambot/Bot.py [hunk 15]
```diff
@@ -0,0 +1,422 @@
...
+            j.sal.fs.createDir(rootpath)
```
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/Bot.py]: done

File: ays_bot/telegrambot/Bot.py [hunk 16]
```diff
@@ -0,0 +1,422 @@
...
+            err_msg = "Error sending message (chat id %s)'%s' : %s" % (data['chat_id'], msg, str(e))
+            self.logger.error(err_msg)
+            resp = {'error': err_msg}
+            self._rediscl.rpush(key, j.data.serializer.json.dumps(resp))
+            return
+
+    def send_event(self, payload):
+        self._rediscl.publish('telegram', payload)
+
+    def reload_cmd(self, bot, update):
```
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/Bot.py]: done

File: ays_bot/telegrambot/Bot.py [hunk 17]
```diff
@@ -0,0 +1,422 @@
...
+            self.logger.debug("call back found")
+            cb = self.question_callbacks.pop(username)
+            cb(bot, update)
+            return
+
+        if getattr(update.message, 'document', None):
+            self.blueprint_mgmt.document(bot, update)
+            return
+
+    def help_cmd(self, bot, update):
```
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/Bot.py]: done

File: ays_bot/telegrambot/Blueprint.py [hunk 18]
```diff
@@ -0,0 +1,424 @@
...
+            msg = 'Error during blueprint execution, check validity of your blueprint.\n Error: %s' % str(e)
+            self.logger.error(e.message)
+        else:
+            msg = "Blueprint deployed. Run with `/run create` then `/run exec`"
+
+        self.bot.sendMessage(
+            chat_id=chat_id,
+            text=msg)
+
+    def list(self, bot, update):
```
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/Blueprint.py]: done

File: ays_bot/telegrambot/Blueprint.py [hunk 19]
```diff
@@ -0,0 +1,424 @@
...
+
+        bp.enable()
+        bot.sendMessage(text="blueprint enabled successfuly",
+                        chat_id=chat_id)
+
+    def delete(self, bot, update, names):
+        username = update.message.from_user.username
+        chat_id = update.message.chat_id
+        bot.sendChatAction(chat_id=chat_id, action=telegram.ChatAction.TYPING)
+        repo = j.atyourservice.repoGet(j.sal.fs.joinPaths(AYS_REPO_DIR, self._currentRepoName(username)))
```
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/Blueprint.py]: done

File: ays_bot/telegrambot/Blueprint.py [hunk 20]
```diff
@@ -0,0 +1,424 @@
...
+            self.disable(bot, update, update.message.text)
+        self.callbacks[username] = cb
+
+        custom_keyboard = [bluelist]
+        reply_markup = telegram.ReplyKeyboardMarkup(custom_keyboard, resize_keyboard=True, one_time_keyboard=True)
+        return self.bot.sendMessage(chat_id=update.message.chat_id,
+                                    text="Which blueprint do you want to disable ?", reply_markup=reply_markup)
+
+    def enable_prompt(self, bot, update):
+        import ipdb; ipdb.set_trace()
```
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/Blueprint.py]: done

File: ays_bot/telegrambot/Repo.py (hunk 3 above)
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/Repo.py]: done

File: ays_bot/telegrambot/Repo.py (hunk 4 above)
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/Repo.py]: done

File: ays_bot/telegrambot/Repo.py (hunk 5 above)
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/Repo.py]: done

File: ays_bot/telegrambot/Service.py (hunk 7 above)
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/Service.py]: done?

File: ays_bot/telegrambot/Service.py (hunk 8 above)
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/Service.py]: done

File: ays_bot/telegrambot/Service.py (hunk 9 above)
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/Service.py]: done

File: ays_bot/telegrambot/run.py (hunk 10 above)
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/run.py]: done

File: ays_bot/telegrambot/run.py (hunk 11 above)
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/run.py]: run

File: ays_bot/telegrambot/run.py (hunk 12 above)
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/run.py]: done?

File: ays_bot/telegrambot/templates/blueprint (hunk 13 above)
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/templates/blueprint]: done

File: ays_bot/telegrambot/README.md [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+Generate a token via @botfather on Telegram, then use it to launch this bot:
+
+With a python script:
+```python
+from JumpScale import j
+bot = j.atyourservice.telegramBot('1857YYYY:XXXXXXXXXXXX')
+bot.run()
+```
+Using ays CLI:
+`ays bot -token 1857YYYY:XXXXXXXXXXXX`
```
Comment (abdulrahmantkhalifa) [inline ays_bot/telegrambot/README.md]: done

**Comment by zaibon** [comment]
merged in 8.1.0
